### PR TITLE
ci: refactor functions to call gcloud

### PR DESCRIPTION
**Pull request by coryan**
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/3773)

### CODE REVIEW

File: ci/kokoro/docker/build-in-docker-bazel.sh [hunk 1]
```diff
@@ -211,15 +146,15 @@ if [[ "${RUN_INTEGRATION_TESTS}" == "yes" || \
...
-  "${GCLOUD}" "${GCLOUD_ARGS[@]}" auth activate-service-account --key-file \
-      "${GOOGLE_APPLICATION_CREDENTIALS}"
+  activate_service_account_keyfile "${GOOGLE_APPLICATION_CREDENTIALS}"
   # This is used in a Bigtable example showing how to use access tokens to
   # create a grpc::Credentials object. Even though the account is deactivated
   # for use by `gcloud` the token remains valid for about 1 hour.
   ACCESS_TOKEN="$("${GCLOUD}" "${GCLOUD_ARGS[@]}" auth print-access-token)"
   readonly ACCESS_TOKEN
+
   # Deactivate the recently activated service accounts to prevent accidents.
```
Comment (devbww) [inline ci/kokoro/docker/build-in-docker-bazel.sh:156]: The comment on the `revoke_service_account_keyfile` command above changed from "Deactivate the recently activated ..." to "Deactivate all ...".  Should this change too?

File: ci/kokoro/docker/build-in-docker-bazel.sh [hunk 2]
```diff
@@ -197,12 +132,12 @@ if [[ "${RUN_INTEGRATION_TESTS}" == "yes" || \
     echo
     echo "================================================================"
     log_normal "Delete service account used in the storage HMAC tests."
```
Comment (devbww) [inline ci/kokoro/docker/build-in-docker-bazel.sh:134]: s/Delete/Activate/?

File: ci/kokoro/docker/download-cache.sh [hunk 3]
```diff
@@ -31,20 +38,30 @@ echo "max_size = 4.0G" >"${HOME_DIR}/.ccache/ccache.conf"
...
   echo "================================================================"
-  echo "Service account for cache access is not configured."
-  echo "No attempt will be made to download the cache, exit with success."
+  log_normal "Service account for cache access is not configured."
+  log_normal "No attempt will be made to download the cache, exit with success."
+  exit 0
+fi
+
+if [[ "${KOKORO_JOB_TYPE:-}" != "PRESUBMIT_GERRIT_ON_BORG" ]] && \
+   [[ "${KOKORO_JOB_TYPE:-}" != "PRESUBMIT_GITHUB" ]]; then
```
Comment (devbww) [inline ci/kokoro/docker/download-cache.sh:47]: Q. Is it worth just matching on `PRESUBMIT_.*`?

File: ci/kokoro/docker/build-in-docker-cmake.sh [hunk 4]
```diff
@@ -205,25 +205,21 @@ if [[ "${BUILD_TESTING:-}" = "yes" ]]; then
...
-    # Extract the service account name so we can deactivate it.
-    GOOGLE_APPLICATION_CREDENTIALS_ACCOUNT="$(sed -n \
-        's/.*"client_email": "\(.*\)",.*/\1/p' \
-        "${GOOGLE_APPLICATION_CREDENTIALS}")"
-    readonly GOOGLE_APPLICATION_CREDENTIALS_ACCOUNT
+    readonly GOOGLE_CLOUD_CPP_BIGTABLE_TEST_ACCESS_TOKEN
 
     # Deactivate the recently activated service account to prevent accidents.
-    "${GCLOUD}" --quiet auth revoke "${GOOGLE_APPLICATION_CREDENTIALS_ACCOUNT}"
+    revoke_service_account_keyfile "${GOOGLE_APPLICATION_CREDENTIALS}"
```
Comment (devbww) [inline ci/kokoro/docker/build-in-docker-cmake.sh:222]: Should "revoke"s also be a `trap EXIT`?

File: ci/kokoro/gcloud-functions.sh [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+  "${GCLOUD}" "${GCLOUD_ARGS[@]}" auth revoke "${account}" >/dev/null
+}
+
+delete_gcloud_config() {
+  "${GCLOUD}" --quiet config configurations delete "${GCLOUD_CONFIG}"
+}
+
+create_gcloud_config() {
+  echo
+  echo "================================================================"
```
Comment (devbww) [inline ci/kokoro/gcloud-functions.sh:50]: Should these be within the `if`?

File: ci/kokoro/gcloud-functions.sh [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+set -eu
+
+if [[ -z "${GCLOUD_CONFIG:-}" ]]; then
+  readonly GCLOUD_CONFIG="cloud-cpp-integration"
+  readonly GCLOUD_ARGS=(
+      # Do not seek confirmation for any actions, assume the default
+      "--quiet"
```
Comment (devbww) [inline ci/kokoro/gcloud-functions.sh:22]: Should the `GCLOUD_ARGS=( '--quiet' )` be outside the `${GCLOUD_CONFIG}` guard?

Even if not, should `gcloud-functions.sh` always be setting some value for `${GCLOUD_ARGS}`?

Comment (coryan) [reply ci/kokoro/gcloud-functions.sh:25]: Fixed, I think, but let me know if I missed the point.

File: ci/kokoro/gcloud-functions.sh (hunk 5 above)
Comment (coryan) [reply ci/kokoro/gcloud-functions.sh]: Yes, fixed.

File: ci/kokoro/docker/build-in-docker-cmake.sh (hunk 4 above)
Comment (coryan) [reply ci/kokoro/docker/build-in-docker-cmake.sh]: Done, but maybe it is a bit awkward. Open to suggestions.

File: ci/kokoro/docker/download-cache.sh (hunk 3 above)
Comment (coryan) [reply ci/kokoro/docker/download-cache.sh:47]: I would prefer to not guess what the current and future job types mean.

File: ci/kokoro/docker/build-in-docker-bazel.sh (hunk 2 above)
Comment (coryan) [reply ci/kokoro/docker/build-in-docker-bazel.sh]: Both actually, created additional log lines to make this clearer.

File: ci/kokoro/docker/build-in-docker-bazel.sh (hunk 1 above)
Comment (coryan) [reply ci/kokoro/docker/build-in-docker-bazel.sh:157]: Actually the one above is wrong. Fixed it.

File: ci/kokoro/gcloud-functions.sh (hunk 6 above)
Comment (devbww) [reply ci/kokoro/gcloud-functions.sh:25]: I was thinking you might just want something along the lines of ...
```
if [[ -z "${GCLOUD_CONFIG:-}" ]]; then
  GCLOUD_CONFIG="cloud-cpp-integration"
fi
readonly GCLOUD_CONFIG

GCLOUD_ARGS+=( "--quiet" "--configuration=${GCLOUD_CONFIG}" )
readonly GCLOUD_ARGS
```

Comment (coryan) [reply ci/kokoro/gcloud-functions.sh:25]: I think the `GCLOUD_ARGS+=` would fail if you source the file twice?

Comment (devbww) [reply ci/kokoro/gcloud-functions.sh:25]: I see ... the `-z` tests are not to allow caller overrides, but rather to allow repeated `source`ing.  OK.  Thanks.

File: ci/kokoro/docker/download-cache.sh [hunk 7]
```diff
@@ -31,20 +38,30 @@ echo "max_size = 4.0G" >"${HOME_DIR}/.ccache/ccache.conf"
...
+create_gcloud_config
+activate_service_account_keyfile "${KEYFILE}"
 gsutil -q cp "gs://${CACHE_FOLDER}/${CACHE_NAME}.tar.gz" "${HOME_DIR}"
-gcloud --quiet auth revoke --all || echo "Ignore revoke failure"
+
 # Ignore timestamp warnings, Bazel has files with timestamps 10 years
 # into the future :shrug:
 tar -zxf "${HOME_DIR}/${CACHE_NAME}.tar.gz" 2>&1 | grep -E -v 'tar:.*in the future'
 
+revoke_service_account_keyfile "${KEYFILE}"
```
[the remark this answers is not in the document]
Comment (coryan) [reply ci/kokoro/docker/download-cache.sh]: Done, with some additional cleanups to all these programs to make them similar.

File: ci/kokoro/docker/upload-cache.sh [hunk 8]
```diff
@@ -20,33 +20,42 @@ if [[ $# != 3 ]]; then
...
 tar -zcf "${HOME_DIR}/${CACHE_NAME}.tar.gz" \
     "${HOME_DIR}/.cache" "${HOME_DIR}/.ccache"
-gcloud --quiet auth activate-service-account --key-file "${KEYFILE}"
+
+trap delete_gcloud_config EXIT
+create_gcloud_config
+activate_service_account_keyfile "${KEYFILE}"
 gsutil -q cp "${HOME_DIR}/${CACHE_NAME}.tar.gz" "gs://${CACHE_FOLDER}/"
-gcloud --quiet auth revoke --all >/dev/null 2>&1 || echo "Ignore revoke failure"
+revoke_service_account_keyfile "${KEYFILE}"
```
[the remark this answers is not in the document]
Comment (coryan) [reply ci/kokoro/docker/upload-cache.sh]: Done.

File: ci/kokoro/macos/download-cache.sh [hunk 9]
```diff
@@ -64,7 +70,6 @@ echo "$(date -u): Extracting build cache ${CACHE_NAME}"
 tar -C / -zxf "${DOWNLOAD}/${CACHE_NAME}.tar.gz" 2>&1 | \
     grep -E -v 'tar:.*in the future'
 
-echo "================================================================"
-echo "$(date -u): build cache extracted"
+revoke_service_account_keyfile "${KEYFILE}"
```
[the remark this answers is not in the document]
Comment (coryan) [reply ci/kokoro/macos/download-cache.sh]: Done.

File: ci/kokoro/macos/upload-cache.sh [hunk 10]
```diff
@@ -59,19 +66,18 @@ readonly UPLOAD="cmake-out/upload"
...
 gsutil -q cp "${UPLOAD}/${CACHE_NAME}.tar.gz" "gs://${CACHE_FOLDER}/"
 
 echo "================================================================"
-echo "$(date -u): Upload completed"
-ACCOUNT="$(sed -n 's/.*"client_email": "\(.*\)",.*/\1/p' "${KEYFILE}")"
-readonly ACCOUNT
-gcloud --quiet auth revoke "${ACCOUNT}" >/dev/null 2>&1 || \
-    echo "Ignore revoke failure"
+log_normal "Upload completed"
+revoke_service_account_keyfile "${KEYFILE}"
```
[the remark this answers is not in the document]
Comment (coryan) [reply ci/kokoro/macos/upload-cache.sh]: Done.

**Comment by coryan** [comment]
PTAL

File: ci/kokoro/docker/build-in-docker-bazel.sh [hunk 11]
```diff
@@ -211,15 +147,15 @@ if [[ "${RUN_INTEGRATION_TESTS}" == "yes" || \
...
+  activate_service_account_keyfile "${GOOGLE_APPLICATION_CREDENTIALS}"
   # This is used in a Bigtable example showing how to use access tokens to
   # create a grpc::Credentials object. Even though the account is deactivated
   # for use by `gcloud` the token remains valid for about 1 hour.
   ACCESS_TOKEN="$("${GCLOUD}" "${GCLOUD_ARGS[@]}" auth print-access-token)"
   readonly ACCESS_TOKEN
+
   # Deactivate the recently activated service accounts to prevent accidents.
-  "${GCLOUD}" "${GCLOUD_ARGS[@]}" auth revoke --all
+  revoke_service_account_keyfile "${GOOGLE_APPLICATION_CREDENTIALS}"
```
Comment (devbww) [inline ci/kokoro/docker/build-in-docker-bazel.sh:158]: Perhaps this one now belongs in the `EXIT` handler too?

Comment (coryan) [reply ci/kokoro/docker/build-in-docker-bazel.sh:171]: Done.

**Pull request merged**